File: fixtures/rename/tests/test_rename.py
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.

import unittest
from uniffi_fixture_rename import *

class TestRename(unittest.TestCase):
    def test_rename(self):
        # Test renamed record
        record = RenamedRecord(renamed_field=42)
        self.assertEqual(record.renamed_field, 42)

        # Test renamed enum
        enum1 = RenamedEnum.RENAMED_VARIANT()
        enum2 = RenamedEnum.RECORD(record)
        self.assertEqual(enum2[0].renamed_field, 42)

        # Test renamed function
        result = renamed_function(record)
        self.assertEqual(result[0].renamed_field, 42)

        # Test renamed object with renamed constructor and method
        obj = RenamedObject.renamed_constructor(123)
        self.assertIsInstance(obj, RenamedObject)
        self.assertEqual(obj.renamed_method(), 123)

        # Test renamed error type exists
        self.assertTrue(hasattr(RenamedError, 'RenamedErrorVariant'))

        # Test trait method renaming.
        trait_impl = create_trait_impl(5)
        self.assertEqual(trait_impl.renamed_trait_method(10), 50) # 10 * 5

    def test_binding_renames(self):
        # Test Python-specific renames for "binding" items
        record = PyRecord(python_item=42)
        self.assertEqual(record.python_item, 42)

        # Test renamed enum
        enum1 = PyEnum.PYTHON_VARIANT_A()
        enum2 = PyEnum.PYTHON_RECORD(record)
        self.assertEqual(enum2[0].python_item, 42)

        # Test renamed enum fields
        enum1 = PyEnumWithFields.PYTHON_VARIANT_A(python_int = 1)
        self.assertEqual(enum1.python_int, 1)
        enum2 = PyEnumWithFields.PYTHON_RECORD(python_record = record, python_int = 1)
        self.assertEqual(enum2.python_record, record)

        # Test renamed function
        result = py_function(python_record=record)
        self.assertEqual(result[0].python_item, 42)

        # Test renamed object with renamed method
        obj = PyObject(python_value = 5)
        self.assertEqual(obj.python_method(python_arg=10), 15)

        # Test renamed error type exists
        self.assertTrue(hasattr(PyError, 'PythonSimple'))

        # Test trait method renaming
        trait = create_binding_trait_impl(5)
        self.assertEqual(trait.python_trait_method(10), 50)

if __name__ == '__main__':
    unittest.main()

File: fixtures/rename/tests/test_rename.swift
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import Foundation
import uniffi_fixture_rename

// Test renamed types - should be accessible by their custom names
let record = RenamedRecord(renamedField: 42)
assert(record.renamedField == 42)

let enum1 = RenamedEnum.renamedVariant
let enum2 = RenamedEnum.record(record)

// Test renamed function
let result = renamedFunction(record: record)
if case .record(let recordResult) = result {
    assert(recordResult.renamedField == 42)
}

// Test renamed object with renamed constructor and method
let obj = RenamedObject.renamedConstructor(value: 123)
let methodResult = obj.renamedMethod()
assert(methodResult == 123)

// Test trait method renaming (trait itself keeps original name)
let traitImpl = createTraitImpl(multiplier: 5)
assert(traitImpl.renamedTraitMethod(value: 10) == 50) // 10 * 5

// Test TOML-based renaming for Swift
// These use the Swift-specific renames from uniffi.toml

// Test renamed record via TOML
let swiftRecord = SwiftRecord(swiftItem: 100)
assert(swiftRecord.swiftItem == 100)

// Test renamed enum via TOML
let swiftEnum1 = SwiftEnum.swiftVariantA
let swiftEnum2 = SwiftEnum.swiftRecord(swiftRecord)
let swiftEnum3 = SwiftEnumWithFields.swiftVariantA(swiftInt: 1)

// Test renamed function with renamed argument via TOML
do {
    let swiftResult = try swiftFunction(swiftRecord: swiftRecord)
    if case .swiftRecord(let recordResult) = swiftResult {
        assert(recordResult.swiftItem == 100)
    }
} catch let error as SwiftError {
    if case .SwiftSimple = error {
        // Expected for nil input
    }
}

// Test that we can also get an error from the function
do {
    let _ = try swiftFunction(swiftRecord: nil)
    fatalError("Should have thrown an error")
} catch let error as SwiftError {
    if case .SwiftSimple = error {
        // Expected
    } else {
        fatalError("Wrong error type")
    }
}

// Test renamed object with renamed method and argument via TOML
let swiftObj = SwiftObject(value: 200)
do {
    let swiftMethodResult = try swiftObj.swiftMethod(swiftArg: 50)
    assert(swiftMethodResult == 250) // 200 + 50
} catch {
    fatalError("Should not have thrown an error")
}

// Test renamed trait with renamed method via TOML
let swiftTraitImpl = createBindingTraitImpl(multiplier: 3)
assert(swiftTraitImpl.swiftTraitMethod(value: 7) == 21) // 7 * 3
